### PR TITLE
PLA-225 Add a rule banning all type assertions

### PR DESCRIPTION
**Pull request by quesabe**
Closes PLA-225.

The rule completely disallows type assertions (except `as const`). It can be silenced with an `eslint-disable` comment which might be supplemented with `eslint-comments/require-description` rule in order to require a comment on each type assertion.

### CODE REVIEW

File: tests/lib/utils/tsconfig.json
Comment (gvidon) [inline tests/lib/utils/tsconfig.json:1]: @quesabe what are we using utils for?

File: lib/rules/ts-as-expression.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      description: 'Disallows the use of Type Assertions within TypeScript files',
+      recommended: false,
+      url: 'https://github.com/ottofeller/eslint-plugin-ottofeller.git'
+    },
+
+    messages: {noTypeAssertions: 'Type Assertions are not allowed'},
+    type: 'suggestion',
+  },
+
+  create(context) {
```
Comment (gvidon) [inline lib/rules/ts-as-expression.js:26]: @quesabe aren’t there ready to go plugins available?

**Comment by Serator** [comment]
@quesabe https://typescript-eslint.io/rules/consistent-type-assertions/#:~:text=Some%20codebases%20like%20to%20go%20for%20an%20extra%20level%20of%20type%20safety%2C%20and%20ban%20assertions%20altogether%20via%20the%20never%20option. - doesn't `consistent-type-assertions` with the `assertionStyle: never` option do the task?

**Comment by Serator** [comment]
[https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=PTAEDUEMBsEsBMBcBYAUAbzaUsC2AHAewCcAXUdUAFQGUBRGq4gU2dAF9QAzYw3UAOQABUgE98zAM4BjYrHykAtFLgA7UsACupWNEkCsoeM2nRILUNMKrJ5XM1wAjZsUTV6jFswB0tOqs1cAFkHZ2JDWC5QAApDbHsnF28EbzEJUABeLNAAQUYAfQA5AHkAETp8qgBNAAUGbwAZWFIXGFAAMna40DTmQiiEsOT4bwA3GE02LIzBWzlVAHMDVGwASgpuqxtyVUh7TNBBpIRDdjQzjEMt21BIA4BtAEYAGlAAJgBdK+sbxweAZleABYvitLD9yNIHt4YZBXjDvI4PrdJODtuc0JgwddyHcZk9Xp8UWjbBjLtiIbcDgJIAJiTiyVjsDiqTNKFxCIQ3AIOYQ6ZxIKiGagLkySbiHi93qDmZS-vjAaAQd9tpYDgAeHEAPnuCLhoARSMZKpueNAmohOqln2NFNVZot2y1NOWopNEpmjtsWvZnO5vP5ZLQIFAAElVOM4EhMe7uJyDtF-qAANSgACs60FoACiXCIpjdpuvIOSazObCttlquLnvLLi1-0r4rjhA1-wbTZZNdApdR-wA3J3Kd31XXiFrEyn06sh9X4zNJ6mM8TIKpRLObpHJtSWrY6Vm5rBFqAAD7ZzTQaCn0CaVTGLhH5jwJvRGiidSQAAeK7Xq28OWIYhIFEOhP3wFhJEkWBrGpAMm16UAADF53PXNY27Rdp2JZDCHg8Q2Bwg4x3QlD1RwickyXGd83JbBjFMcw2C7P0kM5WNHHMDVyN5PD0ioTR8GgKZQHuRUBAAC1gARXgERxRGYAQZWbbtRJkyTpMEOSFORLN+ME5heLYGpCEg2BHCEvShIOShxOYS8uUEWz7MDQtyG7Gy7OgByJM8vkOGJYzTPM5hLIMmixQQwKoOC0LrNAAAGbknK8jS0yS3yXKrIsUMoRLHIy140vy5z-KzKKzIsgShKbPAiDICh3AYJhWH8ng+EEER8JkOQFGUSQ1A0bRdH0Qx6LMCwWSOVxGs8VhfHoHMQjQ1zsz2YSpuGYk-FmnwaFIeYFiaFogOgc4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra0ipWgENkyRNHyUuqDAG1IU6B2iQANJkjjJ02UwDKRJOihNEANymQAvgF1wYWyFtA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA - ](https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=PTAEDUEMBsEsBMBcBYAUCUBvNHSlgLYAOA9gE4AuWoAKgMoCidNZApq6AL6gBmZJBUAHIAAhQCeRVgGcAxmVhEKAWhlwAdhWABXCrGjShOMHnitZ0SG1CyS66VQKsCAI1ZlEtRszasAdPQM6toEALLObmTGePg8oAAU0TGgTq7ufgh+ElKgALz5oACCzAD6AHIA8gAiDCU0AJoACkx+ADKwFO4woABkPUkx2awkcamRGfB+AG4w2hz5ucIOCuoA5kboJqAAlFgDeLb2VOqQTnkpEekIA5zRt6jYqAd2DqCQ5wDaAIwANKAATABdNDPI6gFyfADMfwALMCnjYXlRZJ8-GjIH80X4XIC3tJEUc0PdHqDXu9Ft8-kC8QSHES0CTaVRycJIEIaYc6ahiSCmW9zpheCQSJ4hDxhezuJB8ZyKPSHrzZfyKb8AfDSVQIRToaA4YqkTZzgAeWUAPg+WIxoCxOPljKVLJNSPNqqBdv1YMdZqEbPdCIdxrNgvFIuEIcl8uiAEl1DM4EgGR7XiHzvFIaAANSgACsu2loGCaSi3MT-oNKcW6fzhcifo1QpIxpr7lNkLrfIroCNkNb7aVnar+MhAG4++XhU2QpFTWnMzntmOwZ3Z1nczTIOpxIvXnG5uchJ0HOz88tYGtQAAfAvaaDQS+gbTqMw8M+seDyjCPXAJOjiTSQAAPddN22PxCjIMhIHEBgAKINhpGkWA7H3cM7lLPAhlAAAxCdFmbYt62XdNVzzfEcJIdtMPI858KTKhOyNciZ2I+cPzAL8tjMCwrA4fthU8ciBiVFwrGNJiQzQhUEUwmhtCIaB5lAD4dSEAALWAhD+IQXHEVghHVDtcKUlT1M04QdL03F81k+TWEoyQOEaEgENgFwFJshSBVAVTWFvUM1N86ASAjMslyMwUfL80VIqCyUaScly3NYDy7JLKSMIc0AEsQpKUq8gAGaLAuCv5syKvyQsI8LQEK4QYpKnNytirh4ucnL3LkhS2L2TYYkIUhKGoQIfHYFq+AEYQxAcuQFCUVRpA0LRdH0QwBi4yxrCVMZ3E8YaWHYAJGELcIiyEg0TjORZtrICYaT23w-DoCgVlWdpOkg6A7iAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra0ipWgENkyRNHyUuqDAG1IU6B2iQANJkjjJ02UwDKRJOihNEANymQAvgF1wYWyFtA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) - demo of a rule based on examples from this Pr.

**Comment by quesabe** [comment]
> @quesabe https://typescript-eslint.io/rules/consistent-type-assertions/#:~:text=Some%20codebases%20like%20to%20go%20for%20an%20extra%20level%20of%20type%20safety%2C%20and%20ban%20assertions%20altogether%20via%20the%20never%20option. - doesn't `consistent-type-assertions` with the `assertionStyle: never` option do the task?

Actually I used this plugin as a basis, but failed to mention that this rule has the `assertionStyle: never` option. It surely seems to do the job.

File: tests/lib/utils/tsconfig.json
Comment (quesabe) [reply tests/lib/utils/tsconfig.json:1]: That's for the tests. However with the `consistent-type-assertions` the whole PR deems unnecessary.

Comment (gvidon) [reply tests/lib/utils/tsconfig.json:1]: @quesabe do we close the PR then?

**Pull request closed without merging**